modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PageHandler.java

```diff
@@ -0,0 +1,478 @@
+                init = pageMem.ioRegistry().resolve(pageAddr);
+            }
+
+            R res = h.run(groupId, pageId, page, pageAddr, init, arg, intArg, statHolder);
+
+            ok = true;
+
+            return res;
+        } finally {
+            assert PageIo.getCrc(pageAddr) == 0;
+
+            if (h.releaseAfterWrite(groupId, pageId, page, pageAddr, arg, intArg)) {
+                writeUnlock(pageMem, groupId, pageId, page, pageAddr, lsnr, ok);
+            }
+        }
+    }
+
+    /**
+     * Acquires the write lock on the page.
+     *
+     * @param pageMem Page memory.
+     * @param groupId Group ID.
+     * @param pageId  Page ID.
+     * @param page    Page pointer.
+     * @param lsnr    Lock listener.
+     * @param tryLock Only try to lock without waiting.
+     * @return Page address or {@code 0} if failed to lock due to recycling.
+     */
+    public static long writeLock(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long page,
+            PageLockListener lsnr,
+            boolean tryLock
+    ) {
+        lsnr.onBeforeWriteLock(groupId, pageId, page);
+
+        long pageAddr = tryLock ? pageMem.tryWriteLock(groupId, pageId, page) : pageMem.writeLock(groupId, pageId, page);
+
+        lsnr.onWriteLock(groupId, pageId, page, pageAddr);
+
+        return pageAddr;
+    }
+
+    /**
+     * Releases acquired write lock.
+     *
+     * @param pageMem  Page memory.
+     * @param groupId  Group ID.
+     * @param pageId   Page ID.
+     * @param page     Page pointer.
+     * @param pageAddr Page address.
+     * @param lsnr     Lock listener.
+     * @param dirty    Page is dirty.
+     */
+    public static void writeUnlock(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long page,
+            long pageAddr,
+            PageLockListener lsnr,
+            boolean dirty
+    ) {
+        lsnr.onWriteUnlock(groupId, pageId, page, pageAddr);
+
+        pageMem.writeUnlock(groupId, pageId, page, dirty);
+    }
+
+    /**
+     * Invokes {@link PageIo#initNewPage(long, long, int)} and does additional checks.
+     */
+    private static void doInitPage(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long pageAddr,
+            PageIo init
+    ) throws IgniteInternalCheckedException {
+        assert PageIo.getCrc(pageAddr) == 0;
+
+        init.initNewPage(pageAddr, pageId, pageMem.realPageSize(groupId));
+    }
+
+    /**
+     * Copies memory from one buffer to another.
+     *
+     * @param src    Source.
+     * @param srcOff Source offset in bytes.
+     * @param dst    Destination.
+     * @param dstOff Destination offset in bytes.
+     * @param cnt    Bytes count to copy.
+     */
+    public static void copyMemory(ByteBuffer src, long srcOff, ByteBuffer dst, long dstOff, long cnt) {
+        byte[] srcArr = src.hasArray() ? src.array() : null;
```

Comment:
Why is this method here? It does not seem belong to 'page handling', it's more about copying. I suggest to create a separate util class and move the method there.

Same applies to the next method.

Reply:
I'll move them to PageUtils.